lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+
+  // Note that this is not bullet-proof. E.g., broken IR can actually increment
+  // dim0 while dim0 already has modulo, since Triton offsets are element-wise
+  // and not in unit of lower dimensions. However, this is highly unlikely but
+  // the analysis will provide wrong result. Hence we provide a warning in this
+  // case.
+  PtrSState const *lhs = &lhsState;
+  PtrSState const *rhs = &rhsState;
+
+  if (rhs->hasModulo()) {
+    std::swap(lhs, rhs);
+  }
+
+  for (uint64_t i = 0; i < lhs->getRank(); i++) {
+    if (!lhs->dimHasModulo(i)) {
+      modulos.push_back(lhs->modulos[i]);
+    } else if (hasConstZero(rhs->offsets[i])) {
+      modulos.push_back(lhs->modulos[i]);
+    } else if (i == 0 && lhs->getRank() == 2 && rhs->scalar) {
+      modulos.push_back(lhs->modulos[1]);
+      modulos.push_back(lhs->modulos[0]);
+      op->emitWarning(
+          "PtrAnalysis: allowing adding pointer state with modulo in dim 0 to "
+          "another pointer state with offset in dim 0.\nPlease verify the "
+          "operand that contains a scalar is meant to increment pointers in "
+          "dim1. If that is not the case it WILL LEAD TO WRONG COMPILATION "
+          "RESULTS.\n\nTo avoid this warning, use expand_dims (instead of "
+          "splat) to explicitly specify which dimension contains the scalar.");
+      break;
+    } else {
+      op->emitRemark(
+          "PtrAnalysis: do not support adding to operand with modulo");
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+LogicalResult PtrSState::mulState(const PtrSState &lhsState,
+                                 const PtrSState &rhsState, Operation *op,
+                                 OpBuilder &builder) {
+  assert(isEmpty() && lhsState.getRank() == rhsState.getRank());
+
+  auto loc = op->getLoc();
+
+  // neither lhs nor rhs should have source, since multiplying base pointer
+  // does not make sense
+  if (lhsState.source && rhsState.source) {
+    op->emitRemark("PtrAnalysis: do not support multiplying base pointer");
+    return failure();
+  }
+
+  assert(!(lhsState.scalar && rhsState.scalar) &&
+         "do not expect to see both lhs and rhs are scalars");
+
+  // currently do not support both tensors are effectively non-scalar
+  if (!lhsState.scalar && !rhsState.scalar) {
+    op->emitRemark(
+        "PtrAnalysis: only support multiplying pointer states when one of "
+        "them represent a scalar");
+    return failure();
+  }
+
+  PtrSState const *lhs = &lhsState;
+  PtrSState const *rhs = &rhsState;
+
+  if (!rhs->scalar && lhs->scalar) {
+    std::swap(lhs, rhs);
+  }
+
+  for (uint64_t i = 0; i < lhs->sizes.size(); i++) {
+    OpFoldResult newOffset =
+        mulOFRValue(lhs->offsets[i], rhs->scalar, loc, builder);
+    OpFoldResult newStride =
+        mulOFRValue(lhs->strides[i], rhs->scalar, loc, builder);
+    OpFoldResult newModulo =
+        mulOFRValue(lhs->modulos[i], rhs->scalar, loc, builder);
+    offsets.push_back(newOffset);
```

Comment:
We had this tracked in an issue already, but I think this is better if we fail the analysis if we are multiplying an expression containing modulo with a scalar. Let's take the following example:

Assume we have a tensor with `cols = 4`

The user wants to load a `1x3` tensor starting at offset 2 while also modding it with 4

```
(2 + tl.arange(0, 3)) % 4
= (2 + [0, 1, 2]) % 4
= ([2, 3, 4] % 4)
= [2, 3, 0]

   0    1    2    3
0  x         x    x
1
2
3
```

This is all good.

But now the user wants to multiply the offsets by 2:

```
((2 + tl.arange(0, 3)) % 4) * 2
= [2, 3, 0] * 4
= [8, 12, 0] <----- these offsets are completely non-sense
```

These offsets are now all over the place, and us multiplying the mod (`4`) with `2` doesn't make sense either.

Reply:
I think multiplying the parent sizes will produce incorrect results though (or rather cases that our formulas won't work with). With the above example again:

```
((2 + tl.arange(0, 3)) % 4) * 2
= [2, 3, 0] * 4
= [8, 12, 0]
```

We finish analyzing `(2 + tl.arange(0, 3)) % 4)` first, at this point we have:
+ offset = 2
+ parent size = 4
+ stride = 1

Then after processing the multiplication, we get:
+ offset = 4
+ parent size = 8
+ stride = 2

So this is when it gets tricky, first the stride is not 1. Our existing offset calculation assumes that the stride is 1 when wrapping around. Also, we will process this as loading from a tensor starting at offset 4 but may wrap around at column 8, when in fact the result `[8, 12, 0]` doesn't represent that; I don't know what it actually means tbh lol. In any case, I think it might be better if we just fail the analysis.

Reply:
Ah I see your concern.

> Then after processing the multiplication, we get parent size = 8
I should not have multiplied the modulos field. That's on me. Thank you for the catch. If it is not multiplied, and if the starting offset is no wrapped around, I think we are still good.

But I agree with you it does gets complicated and . Will remove it for now.

Reply:
We discussed offline, let's use this approach.